wire/protobuf/serializer.go

```
@@ -0,0 +1,182 @@
+	"io"
+
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"perun.network/go-perun/client"
+	"perun.network/go-perun/wire"
+)
+
+type serializer struct{}
+
+func init() {
+	wire.SetEnvelopeSerializer(serializer{})
+}
+
+// Encode encodes an envelope from the reader using protocol buffers
+// serialization format.
+func (serializer) Encode(w io.Writer, env *wire.Envelope) (err error) { //nolint: funlen, cyclop
+	protoEnv := &Envelope{}
+
+	switch msg := env.Msg.(type) {
+	case *wire.PingMsg:
+		protoEnv.Msg = fromPingMsg(msg)
+	case *wire.PongMsg:
+		protoEnv.Msg = fromPongMsg(msg)
+	case *wire.ShutdownMsg:
+		protoEnv.Msg = fromShutdownMsg(msg)
+	case *wire.AuthResponseMsg:
+		protoEnv.Msg = &Envelope_AuthResponseMsg{}
+	case *client.LedgerChannelProposalMsg:
+		protoEnv.Msg, err = fromLedgerChannelProposalMsg(msg)
+	case *client.SubChannelProposalMsg:
+		protoEnv.Msg, err = fromSubChannelProposalMsg(msg)
+	case *client.VirtualChannelProposalMsg:
+		protoEnv.Msg, err = fromVirtualChannelProposalMsg(msg)
+	case *client.LedgerChannelProposalAccMsg:
+		protoEnv.Msg, err = fromLedgerChannelProposalAccMsg(msg)
+	case *client.SubChannelProposalAccMsg:
+		protoEnv.Msg = fromSubChannelProposalAccMsg(msg)
+	case *client.VirtualChannelProposalAccMsg:
+		protoEnv.Msg, err = fromVirtualChannelProposalAccMsg(msg)
+	case *client.ChannelProposalRejMsg:
+		protoEnv.Msg = fromChannelProposalRejMsg(msg)
+	case *client.ChannelUpdateMsg:
+		protoEnv.Msg, err = fromChannelUpdateMsg(msg)
+	case *client.VirtualChannelFundingProposalMsg:
+		protoEnv.Msg, err = fromVirtualChannelFundingProposalMsg(msg)
+	case *client.VirtualChannelSettlementProposalMsg:
+		protoEnv.Msg, err = fromVirtualChannelSettlementProposalMsg(msg)
+	case *client.ChannelUpdateAccMsg:
+		protoEnv.Msg = fromChannelUpdateAccMsg(msg)
+	case *client.ChannelUpdateRejMsg:
+		protoEnv.Msg = fromChannelUpdateRejMsg(msg)
+	case *client.ChannelSyncMsg:
+		protoEnv.Msg, err = fromChannelSyncMsg(msg)
```

Please throw an error in the default case.

---

Add in this commit: d995e181.